src/modules/bulletproofs/rangeproof_impl.h

```diff
@@ -0,0 +1,792 @@
+        default:
+            VERIFY_CHECK(ctx->count < 5 + ctx->n_commits);
+
+            secp256k1_scalar_mul(sc, &ctx->zsq, &ctx->randomizer61);
+            secp256k1_scalar_mul(&ctx->zsq, &ctx->zsq, &ctx->z);
+            *pt = ctx->commit[ctx->count - 5];
+            break;
+        }
+        secp256k1_scalar_mul(sc, sc, randomizer);
+        ctx->count++;
+    }
+    return 1;
+}
+
+static int secp256k1_bulletproof_rangeproof_verify_impl(const secp256k1_ecmult_context *ecmult_ctx, secp256k1_scratch *scratch, const unsigned char* const* proof, const size_t n_proofs, const size_t plen, size_t nbits, const uint64_t* const* min_value, const secp256k1_ge* const* commitp, size_t n_commits, const secp256k1_ge *value_gen, const secp256k1_bulletproof_generators *gens, const unsigned char* const* extra_commit, size_t *extra_commit_len) {
+    secp256k1_bulletproof_vfy_ecmult_context *ecmult_data;
+    secp256k1_bulletproof_innerproduct_context *innp_ctx;
+    int ret;
+    size_t i;
+    int same_generators = 1;
+
+    /* sanity-check input */
+    if (POPCOUNT(nbits) != 1 || nbits > MAX_NBITS) {
+        return 0;
+    }
+    if (plen < 64 + 128 + 1) {  /* inner product argument will do a more precise check */
+        return 0;
+    }
+    if (plen > SECP256K1_BULLETPROOF_MAX_PROOF) {
+        return 0;
+    }
+
+    if (!secp256k1_scratch_allocate_frame(scratch, n_proofs * (sizeof(*ecmult_data) + sizeof(*innp_ctx)), 2)) {
+        return 0;
+    }
+    ecmult_data = (secp256k1_bulletproof_vfy_ecmult_context *)secp256k1_scratch_alloc(scratch, n_proofs * sizeof(*ecmult_data));
+    innp_ctx = (secp256k1_bulletproof_innerproduct_context *)secp256k1_scratch_alloc(scratch, n_proofs * sizeof(*innp_ctx));
+
+    /* In general you cannot memcmp secp256k1_ge's like this because their field
+     * elements may represent the same number differently. In this case it is ok
+     * because (a) a false positive here is no big deal, it will add one mult per
+     * proof to he giant ecmult_multi at the end but not change any semantics;
+     * and (b) typically this list of generators was deterministically decoded
+     * from a list of secp256k1_generators which have a compact encoding, so that 
+     * equal group elements actually will compare equal. */
+    for (i = 1; i < n_proofs; i++) {
+        if (memcmp(&value_gen[i], &value_gen[i - 1], sizeof(value_gen[i])) != 0) {
+            same_generators = 0;
+        }
+    }
+
+    for (i = 0; i < n_proofs; i++) {
+        secp256k1_sha256 sha256;
+        unsigned char commit[32] = {0};
+        unsigned char randomizer61[32] = {0};  /* randomizer for eq (61) so we can add it to eq (62) to save a separate multiexp */
+        secp256k1_scalar taux, mu;
+        secp256k1_ge age, sge;
+        int overflow;
+        size_t j;
+
+        /* Commit to all input data: min value, pedersen commit, asset generator, extra_commit */
+        if (min_value != NULL && min_value[i] != NULL) {
```

Comment:
This is a branch that influences what will be hashed. This makes me nervous, because different inputs could possibly yield the same hash through different control flow decisions (and I think we don't want to reason about the consequences of such a collision). The obvious way to avoid that is to add prefixes to the hashed values (different constants and counters for the loops) to the hash. For the conditionals, another way is to always hash something. All of this makes the code slightly slower unfortunately...

 (Same issue in lines 202, 215, 229, 223).